### ISSUE TITLE
imapnotify only reports new mail for 5 minutes

### ISSUE THREAD

**Issue by protist** [a-sk/node-imapnotify#8]
Hello again. Unfortunately, issue #5 wasn't entirely resolved. After connecting okay, I only receive notifications for 5 minutes. I then receive no notifications, and then later (fairly consistently, at just over 21 minutes after the initial connection) imapnotify detects an error (`Error registered`) and restarts. Hence, for 16 of 21 minutes, imapnotify isn't working.

I initially worked out how long it takes to timeout by scheduling an email to myself every three minutes. While testing, I feared that I'd perturb the system by "refreshing" the connection from the server's end every three minutes. Despite this, the connection still failed rapidly. I also repeated this experiment with a two minute loop. Finally, I waited just under five minutes, then sent a flurry of emails. Imapnotify is timing out fairly precisely at 5 minutes, I'd say 4:46–5:02.

I ran `DEBUG=1 /usr/bin/imapnotify -c ~/.config/imapnotify/main.js`, but there was no output related to the timeout.

I also wondered if sending `NOOP` would be a fix, but it seems like ~5 minutes is too low for a IMAP server to time me out?

I know that you don't use imapnotify yourself any more, and I appreciate your troubleshooting on previous issues, so please let me know if you didn't want to sort this bug out. (But I'd love it if you did, of course!)

**Comment by a-sk** [comment]
Hi protist, I'm sorry to hear that you have this issue.
I will try to look into this.

**Comment by a-sk** [comment]
Can you try to use this python version of [imapnotify](https://github.com/a-sk/python-imapnotify)?

You will need python >= 3.5
You can install pyimapnotify by running:
`pip install imapnotify`

It understands json config from node-imapnotify:
`pyimapnotify -c ~/.config/imap_inotify/test.json -vvv`

**Comment by protist** [comment]
Yes, it seems to work very well!

I tried pyimapnotify side by side with node-imapnotify. Firstly, node-imapnotify fails almost exactly at 5 minutes. I received a new-email notification 4:58 after I launched it, whereas an email sent at 5:03 failed.

However, pyimapnotify notified me for all emails. I tested pyimapnotify for 36 minutes in total, and it worked perfectly fine!

Thank you so much for this, it looks excellent. I'm very happy to migrate over. However, is it possible to store passwords encrypted? At the moment, I'm using `*.js` with the following line.

```
exports.password = getStdout('python2 -c \'import keyring; print keyring.get_password("mbsync","foo@bar.com")\'');
```

**Comment by a-sk** [comment]
I'm glad it helped. I will add a feature to allow calling external script to get password soon.
Something like "password_eval": "/path/to/script some_args"

**Comment by a-sk** [comment]
By the way, your email server supports NOTIFY capability, it allows to use only _one_ connection to receive notifications about all mailboxes you are interested in. I will look into implementing support for it in pyimapnotify.

**Comment by protist** [comment]
Brilliant! Thank you for all your work! I'll publish an Arch Linux package for pyimapnotify, too.

**Comment by a-sk** [comment]
I released 0.0.2 version of pyimapnotify, it supports password_eval, check it out.

**Comment by protist** [comment]
I can't seem to authenticate. I replaced the `"password": "foo"` line in my config with the following.

```
"password_eval": "python2 -c \"import keyring; print keyring.get_password(\\\"mbsync\\\",\\\"user@domain.com\\\")\"",
```

It _seems_ to get the password properly, but fails to authenticate.

```
$ pyimapnotify -c ~/.config/imapnotify/test.json -vvv
2016-10-08 08:13:03,730 DEBUG [core:84] adding INBOX
2016-10-08 08:13:03,730 INFO [core:70] connecting to domain.com:993 for INBOX
2016-10-08 08:13:05,984 ERROR [core:96] could not authenticate to domain.com:993 using user@domain.com:foo
```

In order to test if the error line was showing the expected format, I tried using the original plain-text password, changing one character so I could get an error message, and the output is the same, i.e.

```
2016-10-08 08:13:05,984 ERROR [core:96] could not authenticate to domain.com:993 using user@domain.com:foot
```

**Comment by a-sk** [comment]
Update and try again, please.
I think that '\n' in the command output was the issue.

**Comment by protist** [comment]
Thank you again. Now it works fine for alphanumeric passwords, but not for those with special characters. I haven't extensively troubleshooted yet (I can if you like), but I suspect it might be `"`?

**Comment by a-sk** [comment]
Strange, and does it print those passwords correctly (in the error message)?
I tested with the account you gave me, I used simple echo "password", but it worked.

**Comment by protist** [comment]
The account I gave to you previously didn't have a `"` in its password. However, I've just modified it. Now it's the same password that I emailed you previously, with a single `"` appended. And now pyimapnotify doesn't work, as above. Presumably an error in escaping the double quote?

Also, when I use an account that _does_ work (i.e. just an alphanumeric password), I get the following error. However, it seems to work correctly, I think.

```
2016-10-09 20:31:57,173 DEBUG [core:135] [INBOX] on_new_message: /home/protist/.local/bin/imap_refresh work:INBOX
2016-10-09 20:31:57,174 DEBUG [core:153] [['/home/protist/.local/bin/imap_refresh', 'work:INBOX']] firing callback <function run_command at 0x7f84a4159268>
2016-10-09 20:31:58,892 DEBUG [core:156] [['/home/protist/.local/bin/imap_refresh', 'work:INBOX']] callback result: bytearray(b'\x1b[35mRunning notmuch new\x1b[0m\nProcessed 60 total files in almost no time.\x1b[K\nAdded 2 new messages to the database. Detected 2 file renames.\n')
2016-10-09 20:31:58,892 DEBUG [core:145] [INBOX] on_new_message_post: 
2016-10-09 20:31:58,893 DEBUG [core:153] [[]] firing callback <function run_command at 0x7f84a4159268>
2016-10-09 20:31:58,893 ERROR [core:160] [[]] error during running callback <function run_command at 0x7f84a4159268>: create_subprocess_exec() missing 1 required positional argument: 'program'
```

**Comment by a-sk** [comment]
Yeah, you were right, `"` needed escaping.
Try again, please.

**Comment by protist** [comment]
Works perfectly! Thank you! Also, the callback error above seems to have been fixed.

I'll look to implementing a systemd.service file for the AUR package, too. (I uploaded [python-imapnotify](https://aur.archlinux.org/packages/python-imapnotify/) a few days ago.)

I'm not sure what you want to do with this issue, i.e. close it? I guess python-imapnotify means that node-imapnotify is dead? Feel free to close this issue as you see fit.
